arcaflow_plugin_pcp/pcp_schema.py

```diff
@@ -21,81 +32,28 @@ class PcpInputParams:
             "Timeout in seconds after which to cancel the pmlogger collection"
         ),
     ] = None
-
-
-@dataclass
-class IntervalOutput:
-    interval: typing.Annotated[
-        int,
-        schema.id("@interval"),
-        schema.name("Interval ID"),
-        schema.description("The interval ID as reported by PCP"),
-    ]
-    timestamp: typing.Annotated[
-        str,
-        schema.id("@timestamp"),
-        schema.name("Interval timestamp"),
-        schema.description("The timestamp of the reported interval"),
-    ]
-    commit: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("commit"),
-        schema.description("The commit for the interval"),
-    ] = None
-    disk: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("disk"),
-        schema.description("The disk structure for the interval"),
-    ] = None
-    kbin: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("kbin"),
-        schema.description("The KB in value for the interval"),
-    ] = None
-    kbout: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("kbout"),
-        schema.description("The KB out value for the interval"),
-    ] = None
-    kernel: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("kernel"),
-        schema.description("The kernel CPU structure for the interval"),
-    ] = None
-    mem: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("mem"),
-        schema.description("The memory structure for the interval"),
-    ] = None
-    memused: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("memused"),
-        schema.description("The memory used value for the interval"),
-    ] = None
-    pktin: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("pktin"),
-        schema.description("The packets in for the interval"),
-    ] = None
-    pktout: typing.Annotated[
-        typing.Optional[typing.Dict[str, typing.Any]],
-        schema.name("pktout"),
-        schema.description("The packets out for the interval"),
+    pmlogger_conf: typing.Annotated[
+        typing.Optional[str],
+        schema.name("pmlogger configuration file"),
+        schema.description(
+            "Complete configuration file content for pmlogger as a multi-line string."
+            " If no config file is provided, a default one will be generated."
+        ),
     ] = None
 
 
-interval_output_schema = schema.ListType(plugin.build_object_schema(IntervalOutput))
-
-
 @dataclass
 class PerfOutput:
     pcp_output: typing.Annotated[
-        typing.List[IntervalOutput],
+        typing.List[typing.Any],
```

Comment:
What types will be in here?

Reply:
The output is always a list of dicts, and there are a couple of consistent key-value pairs in each dict, but then the sub-structure of the dict could have far too many variations to build an explicit schema for. But if I can get more specific here, I'd love some recommendations.

Reply:
Every list item in the output has `@interval` and `@timestamp` keys both with string values, and then an unpredictable number of additional keys with multi-level dictionaries as values. So I suppose at the top level of each list item the only two valid value types are strings and dicts.

Reply:
I tried a few different things here, but I can't come up with an alternative solution to `typing.List[typing.Any]`. Here is example output (from the test code):

```python
{
    "@interval": "0",
    "@timestamp": "2023-11-10T10:12:13.531775Z",
    "kernel": {
        "all": {
            "load": {
                "@instances": [
                    {"name": "1 minute", "value": 1.3},
                    {"name": "5 minute", "value": 1.33},
                    {"name": "15 minute", "value": 1.38},
                ]
            }
        }
    },
    "mem": {
        "util": {"used": {"@unit": "Kbyte", "value": 15401408}}
    },
},
{
    "@interval": "1",
    "@timestamp": "2023-11-10T10:12:14.027241Z",
    "disk": {
        "all": {
            "read": {"@unit": "count/s", "value": 14},
            "write": {"@unit": "count/s", "value": 263.993},
        }
    },
    "kernel": {
        "all": {
            "cpu": {
                "sys": {"@unit": "ms/s", "value": 439.988},
                "user": {"@unit": "ms/s", "value": 759.979},
            },
            "load": {
                "@instances": [
                    {"name": "1 minute", "value": 1.3},
                    {"name": "5 minute", "value": 1.33},
                    {"name": "15 minute", "value": 1.38},
                ]
            },
        }
    },
    "mem": {
        "util": {"used": {"@unit": "Kbyte", "value": 15464512}}
    },
},
```

Reply:
What's an example of how the prior way of doing this was limiting?

Reply:
In the existing implementation, we are explicit about the pcp2json configuration and therefore what parameters are returned in the output, so we always knew the objects to expect. The new design allows the user to both customize the pmlogger configuration, which can change the available metrics, as well as the pcp2json configuration, which changes which of the metrics is reported in the output. I think it would be untenable to define every possible parameter for the output while also allowing this flexibility to the user.

Reply:
What I would like to do here is have the schema expect for each dict item in the list output the `interval` and `timestamp` object explicitly, with validation, and then any number of `any` type objects, but I don't know of a way to do that without modifying the json output structure to create a sub-list object for all of the metrics outputs (kernel, mem, cpu, etc.). That's a possible approach, but I'm not convinced it adds much value.